### DISCUSSION TITLE
Is there way to find out database session id in the connection object when we connect to the database

### DISCUSSION THREAD

**Discussion by sidnaik03** [sqlalchemy/sqlalchemy#6954] [Usage Questions]
### Describe the use case

Generally when we connect to any database, database creates an unique session id against each connection. That session id will still remain with opened connection, until we close that connection. In case of no active query running on database it , it remain as 'INACTIVE' status.
As this feature currently not available (as I believe), I have to fire the separate query to fetch session id after the connection is happen. 

### Databases / Backends / Drivers targeted

I believe this common for all RDBMS database. In my case I am looking for the Oracle, MSSQL and DB2i. 
Drivers I am using 
Oracle - cx_oracle
MSSQL - pyodbc
DB2i - pyodbc

### Example Use

`
# example for oracle and mssql
from sqlalchemy import create_engine

oracle_engine = create_engine('oracle+cx_oracle://TEST:********@localhost:1521/?service_name=E1LOCAL',echo=True)
mssql_engine = create_engine("mssql://test:******@localhost:1433/testdb?driver=ODBC Driver 17 for SQL Server", echo=True)
oracle_conn =  oracle_engine.connect()
mssql_conn = mssql_engine.connect()

oracle_session = conn.execute('select sid from v$mystat where rownum=1').fetchall()
mssql_session = conn.execute('select session_id from sys.dm_exec_sessions').fetchall()
`

### Additional context

This feature would be helpful to find out which application query causing the server performance . DBA can give us the session id and we can find out which query causing the issue or is there optimization we can do. Also it will help to find out is query actually running on the database or it get stuck in between (Application call and database).
In my case, I am opening the connection only once and it get shared my multiple celery workers. Each worker sending his own query. So having session id against connection it would be helpful to logging and tracing.

**Comment by zzzeek** [comment]
these queries are highly backend specific and I dont think it's safe to assume every backend has such a concept represented, such as SQLite.   your use case is highly specific to the database engine you're targeting so I think running the query as you are doing on the connection is the way to go.

**Comment by zzzeek** [comment]
here's how I'd recommend to implement this:

```py
def connection_sid(connection):
    try:
        stmt = {
            "oracle":'select sid from v$mystat where rownum=1',
            "mssql": 'select session_id from sys.dm_exec_sessions'
        }[connection.engine.name]
    except KeyError:
        raise NotImplementedError(f"unsupported backend {connection.engine.name}")
    else:
        return conn.scalar(text(stmt))
```

**Comment by CaselIT** [comment]
Not really in answer to the question

> In my case, I am opening the connection only once and it get shared my multiple celery workers. Each worker sending his own query. So having session id against connection it would be helpful to logging and tracing.

I would suggest against this pattern. You should not share a connection between multiple workers.
I would make the worker get a connection from the engine when starting to execute a task and then closing it when the task has finisced processing
